core/src/main/scala/caliban/wrappers/ApolloPersistedQueries.scala

```diff
@@ -1,58 +1,125 @@
+    def isValidated(hash: String, variablesHashCode: Int): ZIO[ApolloPersistence, Nothing, Boolean]     =
+      ZIO.serviceWithZIO[ApolloPersistence](_.isValidated(hash, variablesHashCode))
+    def registerValidation(hash: String, variablesHashCode: Int): ZIO[ApolloPersistence, Nothing, Unit] =
+      ZIO.serviceWithZIO[ApolloPersistence](_.registerValidation(hash, variablesHashCode))
 
-    val live: UIO[ApolloPersistence] = Ref.make[Map[String, Document]](Map()).map { cache =>
-      new ApolloPersistence {
-        override def get(hash: String): UIO[Option[Document]]      = cache.get.map(_.get(hash))
-        override def add(hash: String, query: Document): UIO[Unit] = cache.update(_.updated(hash, query))
+    val live: UIO[ApolloPersistence] =
+      (ConcurrentMap.empty[String, Document] <*> ConcurrentSet.empty[(String, Int)]).map {
+        case (docCache, validationCache) =>
+          new ApolloPersistence {
+            override def get(hash: String): UIO[Option[Document]]                            = docCache.get(hash)
+            override def add(hash: String, query: Document): UIO[Unit]                       = docCache.put(hash, query).unit
+            override def isValidated(hash: String, variablesHashCode: Int): UIO[Boolean]     =
+              validationCache.contains(hash -> variablesHashCode)
+            override def registerValidation(hash: String, variablesHashCode: Int): UIO[Unit] =
+              validationCache.add(hash -> variablesHashCode).unit
+          }
       }
-    }
   }
 
   val live: Layer[Nothing, ApolloPersistence] = ZLayer(ApolloPersistence.live)
 
-  private def parsingWrapper(docVar: Ref[Option[Either[String, Document]]]): ParsingWrapper[ApolloPersistence] =
+  private def parsingWrapper(
+    docVar: Promise[Nothing, Option[(String, GraphQLRequest, Option[Document])]]
+  ): ParsingWrapper[ApolloPersistence] =
     new ParsingWrapper[ApolloPersistence] {
       override def wrap[R1 <: ApolloPersistence](
         f: String => ZIO[R1, CalibanError.ParsingError, Document]
       ): String => ZIO[R1, CalibanError.ParsingError, Document] =
         (query: String) =>
-          docVar.getAndSet(None).flatMap {
-            case Some(Right(doc)) => ZIO.succeed(doc)
-            case Some(Left(hash)) => f(query).tap(doc => ApolloPersistence.add(hash, doc))
-            case None             => f(query)
+          docVar.await.flatMap {
+            case Some((_, _, Some(doc))) => ZIO.succeed(doc)
+            case Some((hash, _, None))   => f(query).tap(doc => ApolloPersistence.add(hash, doc))
+            case None                    => f(query)
           }
     }
 
+  private def validationWrapper(
+    docVar: Promise[Nothing, Option[(String, GraphQLRequest, Option[Document])]]
+  ): ValidationWrapper[ApolloPersistence] =
+    new ValidationWrapper[ApolloPersistence] {
+      override val priority: Int = Int.MaxValue // Run this wrapper at the outmost boundary
```

Comment:
What if we changed this from `Int.MaxValue` to some arbitrary number (e.g, 100)? As for the ordering, it might work but that requires existing users to be aware of this change. My thinking here was that existing users get the benefit without having to do any changes to their code.

One other benefit of using a number for the priority is for wrappers that have multiple underlying wrappers, but you'd want those to run in different order

Reply:
Changed this to `100` instead, this way users can have a higher priority wrapper if they wish so